app/controllers/facility_user_reservations_controller.rb

```diff
@@ -18,35 +19,39 @@ def index
                      .paginate(page: params[:page])
   end
 
-  # PUT /facilities/:facility_id/users/:user_id/reservations/:id/cancel
+  # PUT /facilities/:facility_id/users/:user_id/reservations/:order_detail_id/cancel
   def cancel
-    order_detail = user_order_details.find(params[:id])
-    raise ActiveRecord::RecordNotFound unless order_detail.reservation.try(:can_cancel?)
-    order_detail.transaction do
-      if cancel_with_fee(order_detail)
-        flash[:notice] = text("cancel.success")
-      else
-        flash[:error] = text("cancel.error")
-        raise ActiveRecord::Rollback
-      end
+    reservation = @order_detail.reservation
+    raise ActiveRecord::RecordNotFound if reservation.blank?
+    unless reservation.canceled?
+      raise ActiveRecord::RecordNotFound unless reservation.can_cancel?
+      cancel_with_fee!
     end
+    flash[:notice] = text("cancel.success")
     redirect_to facility_user_reservations_path(current_facility, @user)
   end
 
   private
 
-  def cancel_with_fee(order_detail)
-    order_detail
-      .cancel_reservation(session_user, admin: true, admin_with_cancel_fee: true)
+  def cancel_with_fee!
+    @order_detail.transaction do
+      unless @order_detail.cancel_reservation(session_user, admin: true, admin_with_cancel_fee: true)
```

Comment:
Is there anything reasonable that would cause this to fail? Seems like if this fails, then you would get both the success and error flashes.